detector/pattern_detector.go

```diff
@@ -2,42 +2,77 @@ package detector
 
 import (
 	"fmt"
-	"talisman/gitrepo"
-
 	log "github.com/Sirupsen/logrus"
+	"sync"
+	"talisman/gitrepo"
 )
 
 type PatternDetector struct {
 	secretsPattern *PatternMatcher
 }
 
+type match struct {
+	name       gitrepo.FileName
+	path       gitrepo.FilePath
+	commits    []string
+	detections []string
+}
+
 //Test tests the contents of the Additions to ensure that they don't look suspicious
 func (detector PatternDetector) Test(additions []gitrepo.Addition, ignoreConfig TalismanRCIgnore, result *DetectionResults) {
 	cc := NewChecksumCompare(additions, ignoreConfig)
+	matches := make(chan match, 512)
+	ignoredFilePaths := make(chan gitrepo.FilePath, 512)
+	waitGroup := &sync.WaitGroup{}
+	waitGroup.Add(len(additions))
 	for _, addition := range additions {
-		if ignoreConfig.Deny(addition, "filecontent") || cc.IsScanNotRequired(addition) {
-			log.WithFields(log.Fields{
-				"filePath": addition.Path,
-			}).Info("Ignoring addition as it was specified to be ignored.")
-			result.Ignore(addition.Path, "filecontent")
-			continue
+		go func(addition gitrepo.Addition) {
+			defer waitGroup.Done()
+			if ignoreConfig.Deny(addition, "filecontent") || cc.IsScanNotRequired(addition) {
+				ignoredFilePaths <- addition.Path
+				return
+			}
+			detections := detector.secretsPattern.check(string(addition.Data))
+			matches <- match{name: addition.Name, path: addition.Path, detections: detections, commits: addition.Commits}
+		}(addition)
+	}
+	go func() {
+		waitGroup.Wait()
+		close(matches)
+		close(ignoredFilePaths)
```

Comment:
Can this go-routine be a defer ?
 Does it have to be invoked right here ? 
Why not as soon as the waitgroup is defined ?

Reply:
@svishwanath-tw Should not be `defer` for sure. Because this is the one which closes the channel and breaks the for loop..... It cannot be invoked as soon as we define waitGroup as we need to add jobs to it. I think it is in the correct place.

Reply:
The for loop looks like it is based on a counter ? I'm not sure how the channel is controlling the loop.